File: Languages/go/system-design-hld/fundamentals/consistent_hashing.go
package main

import (
	"crypto/sha256"
	"encoding/binary"
	"fmt"
	"sort"
	"sync"
)

// Consistent Hashing - Distributes keys across nodes in a ring.
// Adding/removing a node only remaps ~K/N keys (K=keys, N=nodes).
// Virtual nodes ensure even distribution.

type HashRing struct {
	mu           sync.RWMutex
	ring         []uint32          // sorted hashes on the ring
	nodes        map[uint32]string // hash → node name
	virtualNodes int               // virtual nodes per physical node
}

func NewHashRing(virtualNodes int) *HashRing {
	return &HashRing{
		nodes:        make(map[uint32]string),
		virtualNodes: virtualNodes,
	}
}

// hashKey generates a consistent hash for a key
func hashKey(key string) uint32 {
	h := sha256.Sum256([]byte(key))
	return binary.BigEndian.Uint32(h[:4])
}

// AddNode adds a physical node with virtual replicas
func (hr *HashRing) AddNode(node string) {
	hr.mu.Lock()
	defer hr.mu.Unlock()
	for i := 0; i < hr.virtualNodes; i++ {
		virtualKey := fmt.Sprintf("%s#%d", node, i)
		hash := hashKey(virtualKey)
		hr.ring = append(hr.ring, hash)
		hr.nodes[hash] = node
	}
	sort.Slice(hr.ring, func(i, j int) bool { return hr.ring[i] < hr.ring[j] })
}

// RemoveNode removes a physical node and all its virtual replicas
func (hr *HashRing) RemoveNode(node string) {
	hr.mu.Lock()
	defer hr.mu.Unlock()
	var newRing []uint32
	for _, hash := range hr.ring {
		if hr.nodes[hash] != node {
			newRing = append(newRing, hash)
		} else {
			delete(hr.nodes, hash)
		}
	}
	hr.ring = newRing
}

// GetNode finds which node a key is assigned to
func (hr *HashRing) GetNode(key string) string {
	hr.mu.RLock()
	defer hr.mu.RUnlock()
	if len(hr.ring) == 0 {
		return ""
	}
	hash := hashKey(key)
	// Binary search for the first ring position >= hash
	idx := sort.Search(len(hr.ring), func(i int) bool {
		return hr.ring[i] >= hash
	})
	// Wrap around
	if idx == len(hr.ring) {
		idx = 0
	}
	return hr.nodes[hr.ring[idx]]
}

// GetNodes returns N distinct nodes for replication
func (hr *HashRing) GetNodes(key string, count int) []string {
	hr.mu.RLock()
	defer hr.mu.RUnlock()
	if len(hr.ring) == 0 {
		return nil
	}
	hash := hashKey(key)
	idx := sort.Search(len(hr.ring), func(i int) bool {
		return hr.ring[i] >= hash
	})

	seen := make(map[string]bool)
	var result []string
	for i := 0; i < len(hr.ring) && len(result) < count; i++ {
		pos := (idx + i) % len(hr.ring)
		node := hr.nodes[hr.ring[pos]]
		if !seen[node] {
			seen[node] = true
			result = append(result, node)
		}
	}
	return result
}

func main() {
	ring := NewHashRing(150) // 150 virtual nodes per physical node

	// Add nodes
	nodes := []string{"node-1", "node-2", "node-3"}
	for _, n := range nodes {
		ring.AddNode(n)
	}

	// Distribute 1000 keys and count distribution
	distribution := make(map[string]int)
	for i := 0; i < 1000; i++ {
		key := fmt.Sprintf("key-%d", i)
		node := ring.GetNode(key)
		distribution[node]++
	}

	fmt.Println("--- Distribution across 3 nodes (1000 keys) ---")
	for node, count := range distribution {
		fmt.Printf("  %s: %d keys (%.1f%%)\n", node, count, float64(count)/10)
	}

	// Verify all nodes got some keys (with 150 vnodes, should be fairly even)
	for _, node := range nodes {
		if distribution[node] < 200 { // at least 20% each
			fmt.Printf("WARNING: %s only has %d keys (might be uneven)\n", node, distribution[node])
		}
	}
	fmt.Println("PASS: all nodes received keys")

	// Test adding a node - check remapping
	fmt.Println("\n--- Adding node-4 ---")
	before := make(map[string]string)
	for i := 0; i < 1000; i++ {
		key := fmt.Sprintf("key-%d", i)
		before[key] = ring.GetNode(key)
	}

	ring.AddNode("node-4")

	remapped := 0
	newDist := make(map[string]int)
	for i := 0; i < 1000; i++ {
		key := fmt.Sprintf("key-%d", i)
		after := ring.GetNode(key)
		newDist[after]++
		if before[key] != after {
			remapped++
		}
	}

	fmt.Printf("  Remapped: %d/1000 keys (%.1f%%)\n", remapped, float64(remapped)/10)
	fmt.Println("  New distribution:")
	for node, count := range newDist {
		fmt.Printf("    %s: %d keys\n", node, count)
	}

	// Ideally ~25% of keys remap when adding 1 of 4 nodes
	if remapped < 500 { // should be significantly less than 50%
		fmt.Println("PASS: minimal remapping on node addition")
	} else {
		panic(fmt.Sprintf("FAIL: too many remapped: %d", remapped))
	}

	// Test removing a node
	fmt.Println("\n--- Removing node-2 ---")
	ring.RemoveNode("node-2")
	orphanNode := ring.GetNode("test-key")
	if orphanNode == "node-2" {
		panic("FAIL: removed node still serving keys")
	}
	fmt.Println("PASS: removed node no longer serving keys")

	// Test replication
	replicas := ring.GetNodes("important-key", 2)
	fmt.Printf("\n--- Replication for 'important-key': %v ---\n", replicas)
	if len(replicas) == 2 && replicas[0] != replicas[1] {
		fmt.Println("PASS: got 2 distinct replicas")
	} else {
		panic("FAIL: replication")
	}
}

File: Languages/java/system-design-hld/fundamentals/ConsistentHashing.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.*;

/**
 * Consistent Hashing Ring Implementation
 *
 * Consistent hashing distributes keys across nodes in a way that minimizes
 * remapping when nodes are added or removed. Each node gets multiple "virtual nodes"
 * on the ring for more even distribution.
 *
 * Key concepts:
 * - Hash ring: A circular space [0, 2^32) where both nodes and keys are mapped
 * - Virtual nodes: Each physical node maps to multiple positions on the ring
 * - Minimal disruption: Adding/removing a node only affects keys in its immediate range
 *
 * Run: javac ConsistentHashing.java && java -ea ConsistentHashing
 */
public class ConsistentHashing {

    private final TreeMap<Long, String> ring = new TreeMap<>();
    private final Map<String, Set<Long>> nodePositions = new HashMap<>();
    private final int virtualNodes;

    /**
     * @param virtualNodes Number of virtual nodes per physical node.
     *                     Higher values give better distribution but use more memory.
     */
    public ConsistentHashing(int virtualNodes) {
        this.virtualNodes = virtualNodes;
    }

    /**
     * Adds a node to the hash ring with its virtual nodes.
     */
    public void addNode(String node) {
        Set<Long> positions = new HashSet<>();
        for (int i = 0; i < virtualNodes; i++) {
            long hash = hash(node + "#VN" + i);
            ring.put(hash, node);
            positions.add(hash);
        }
        nodePositions.put(node, positions);
    }

    /**
     * Removes a node and all its virtual nodes from the ring.
     */
    public void removeNode(String node) {
        Set<Long> positions = nodePositions.remove(node);
        if (positions != null) {
            for (Long pos : positions) {
                ring.remove(pos);
            }
        }
    }

    /**
     * Gets the node responsible for a given key.
     * Walks clockwise on the ring from the key's hash position.
     */
    public String getNode(String key) {
        if (ring.isEmpty()) {
            return null;
        }
        long hash = hash(key);
        // Find the first node position >= key hash (clockwise walk)
        Map.Entry<Long, String> entry = ring.ceilingEntry(hash);
        // Wrap around to the first entry if we've gone past the end
        if (entry == null) {
            entry = ring.firstEntry();
        }
        return entry.getValue();
    }

    /**
     * Returns the number of physical nodes on the ring.
     */
    public int getNodeCount() {
        return nodePositions.size();
    }

    /**
     * SHA-256 based hash function mapped to a long value.
     * SHA-256 provides excellent distribution properties.
     */
    private long hash(String key) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(key.getBytes(StandardCharsets.UTF_8));
            // Use first 8 bytes to construct a long (unsigned via mask)
            long h = 0;
            for (int i = 0; i < 8; i++) {
                h = (h << 8) | (digest[i] & 0xFF);
            }
            return h;
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    // ──────────────────────────────────────────────────────────────
    // Tests
    // ──────────────────────────────────────────────────────────────

    public static void main(String[] args) {
        testEvenDistribution();
        testMinimalRemapping();
        testEdgeCases();
        System.out.println("\n✓ All consistent hashing tests passed!");
    }

    /**
     * Demonstrates that virtual nodes provide reasonably even distribution.
     */
    private static void testEvenDistribution() {
        System.out.println("=== Test: Even Distribution ===");
        ConsistentHashing ch = new ConsistentHashing(150);

        String[] nodes = {"NodeA", "NodeB", "NodeC", "NodeD"};
        for (String node : nodes) {
            ch.addNode(node);
        }

        // Distribute 10000 keys and count per node
        Map<String, Integer> distribution = new HashMap<>();
        int totalKeys = 10000;
        for (int i = 0; i < totalKeys; i++) {
            String assigned = ch.getNode("key-" + i);
            distribution.merge(assigned, 1, Integer::sum);
        }

        System.out.println("Distribution of " + totalKeys + " keys across " + nodes.length + " nodes:");
        double idealPercentage = 100.0 / nodes.length; // 25% each
        for (String node : nodes) {
            int count = distribution.getOrDefault(node, 0);
            double pct = (count * 100.0) / totalKeys;
            System.out.printf("  %s: %d keys (%.1f%%)%n", node, count, pct);
            // With 150 virtual nodes, each should get roughly 15-35% (generous bounds)
            assert count > 1000 : node + " has too few keys: " + count;
            assert count < 4000 : node + " has too many keys: " + count;
        }
        System.out.println("  ✓ Distribution within acceptable bounds (ideal: " + String.format("%.0f%%", idealPercentage) + " each)\n");
    }

    /**
     * Demonstrates minimal remapping when a node is added or removed.
     * Only keys in the affected range should move.
     */
    private static void testMinimalRemapping() {
        System.out.println("=== Test: Minimal Remapping ===");
        ConsistentHashing ch = new ConsistentHashing(150);

        String[] initialNodes = {"Server1", "Server2", "Server3"};
        for (String node : initialNodes) {
            ch.addNode(node);
        }

        // Record initial assignment
        int totalKeys = 10000;
        Map<String, String> initialAssignment = new HashMap<>();
        for (int i = 0; i < totalKeys; i++) {
            String key = "request-" + i;
            initialAssignment.put(key, ch.getNode(key));
        }

        // Add a new node
        ch.addNode("Server4");
        int remappedOnAdd = 0;
        for (int i = 0; i < totalKeys; i++) {
            String key = "request-" + i;
            if (!initialAssignment.get(key).equals(ch.getNode(key))) {
                remappedOnAdd++;
            }
        }

        double remapPercentAdd = (remappedOnAdd * 100.0) / totalKeys;
        System.out.printf("  After adding Server4: %d/%d keys remapped (%.1f%%)%n",
                remappedOnAdd, totalKeys, remapPercentAdd);
        // Ideal remapping is ~1/N of keys (1/4 = 25%). Allow generous range.
        assert remapPercentAdd < 40 : "Too many keys remapped on add: " + remapPercentAdd + "%";
        assert remapPercentAdd > 10 : "Too few keys remapped on add: " + remapPercentAdd + "%";
        System.out.println("  ✓ Remapping on add is close to ideal (~25%)");

        // Remove a node
        Map<String, String> beforeRemove = new HashMap<>();
        for (int i = 0; i < totalKeys; i++) {
            String key = "request-" + i;
            beforeRemove.put(key, ch.getNode(key));
        }

        ch.removeNode("Server2");
        int remappedOnRemove = 0;
        for (int i = 0; i < totalKeys; i++) {
            String key = "request-" + i;
            if (!beforeRemove.get(key).equals(ch.getNode(key))) {
                remappedOnRemove++;
            }
        }

        double remapPercentRemove = (remappedOnRemove * 100.0) / totalKeys;
        System.out.printf("  After removing Server2: %d/%d keys remapped (%.1f%%)%n",
                remappedOnRemove, totalKeys, remapPercentRemove);
        // Only keys that were on Server2 should move (~25%)
        assert remapPercentRemove < 40 : "Too many keys remapped on remove: " + remapPercentRemove + "%";
        assert remapPercentRemove > 10 : "Too few keys remapped on remove: " + remapPercentRemove + "%";
        System.out.println("  ✓ Remapping on remove is close to ideal (~25%)\n");
    }

    /**
     * Edge cases: empty ring, single node, duplicate add.
     */
    private static void testEdgeCases() {
        System.out.println("=== Test: Edge Cases ===");

        ConsistentHashing ch = new ConsistentHashing(10);

        // Empty ring
        assert ch.getNode("any-key") == null : "Expected null for empty ring";
        System.out.println("  ✓ Empty ring returns null");

        // Single node
        ch.addNode("OnlyNode");
        for (int i = 0; i < 100; i++) {
            assert "OnlyNode".equals(ch.getNode("key-" + i)) : "Single node should handle all keys";
        }
        System.out.println("  ✓ Single node handles all keys");

        // Remove non-existent node (should not crash)
        ch.removeNode("GhostNode");
        assert ch.getNodeCount() == 1 : "Node count should still be 1";
        System.out.println("  ✓ Removing non-existent node is safe");

        // Remove last node
        ch.removeNode("OnlyNode");
        assert ch.getNode("key") == null : "Expected null after removing all nodes";
        System.out.println("  ✓ Removing last node returns to empty state\n");
    }
}
